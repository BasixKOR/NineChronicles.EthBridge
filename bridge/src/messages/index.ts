import { ChatPostMessageArguments } from "@slack/web-api";
import { URL } from "url";
import { join } from "path";

export interface Message {
    render(): Partial<ChatPostMessageArguments>;
}

export function combineUrl(base: string, addition: string): string {
    const [path, query] = addition.split("?");
    const url = new URL(base);
    url.pathname = join(url.pathname, path);
    url.search = `?${query || ""}`;
    return url.toString();
}
