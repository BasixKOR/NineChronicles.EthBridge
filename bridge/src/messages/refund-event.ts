import { ChatPostMessageArguments } from "@slack/web-api";
import { Message, combineUrl } from ".";
import { TxId } from "../types/txid";
import { Address } from "../types/address";
import { Decimal } from "decimal.js";
import { ForceOmit } from "../types/force-omit";

export class RefundEvent implements Message {
    private readonly _sender: Address;
    private readonly _requestTxId: TxId;
    private readonly _requestAmount: Decimal;
    private readonly _refundTxId: TxId;
    private readonly _refundAmount: Decimal;
    private readonly _explorerUrl: string;
    private readonly _reason: string;

    constructor(
        explorerUrl: string,
        sender: Address,
        requestTxId: TxId,
        requestAmount: Decimal,
        refundTxId: TxId,
        refundAmount: Decimal,
        reason: string
    ) {
        this._sender = sender;
        this._refundTxId = refundTxId;
        this._refundAmount = refundAmount;
        this._requestTxId = requestTxId;
        this._requestAmount = requestAmount;
        this._explorerUrl = explorerUrl;
        this._reason = reason;
    }

    render(): ForceOmit<Partial<ChatPostMessageArguments>, "channel"> {
        return {
            text: "NCG refund event occurred.",
            attachments: [
                {
                    author_name: "Bridge Event",
                    color: "#42f5aa",
                    fields: [
                        {
                            title: "Reason",
                            value: this._reason,
                        },
                        {
                            title: "Address",
                            value: this._sender,
                        },
                        {
                            title: "Request transaction",
                            value: combineUrl(
                                this._explorerUrl,
                                `transaction?${this._requestTxId}`
                            ),
                        },
                        {
                            title: "Request Amount",
                            value: this._requestAmount.toString(),
                        },
                        {
                            title: "Refund transaction",
                            value: combineUrl(
                                this._explorerUrl,
                                `transaction?${this._refundTxId}`
                            ),
                        },
                        {
                            title: "Refund Amount",
                            value: this._refundAmount.toString(),
                        },
                    ],
                    fallback: `Refund NCG ${this._refundAmount} in ${this._requestAmount} to ${this._sender}`,
                },
            ],
        };
    }
}
