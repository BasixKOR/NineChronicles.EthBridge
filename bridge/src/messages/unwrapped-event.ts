import { ChatPostMessageArguments } from "@slack/web-api";
import { WrappingEvent } from "./wrapping-event";
import { Address } from "../types/address";
import { TxId } from "../types/txid";

export class UnwrappedEvent extends WrappingEvent {
    private readonly _sender: Address;
    private readonly _recipient: Address;
    private readonly _nineChroniclesTxId: TxId;
    private readonly _ethereumTransactionHash: string;
    private readonly _amount: string;

    constructor(
        explorerUrl: string,
        etherscanUrl: string,
        sender: Address,
        recipient: Address,
        amount: string,
        nineChroniclesTxId: TxId,
        ethereumTransactionHash: string
    ) {
        super(explorerUrl, etherscanUrl);

        this._sender = sender;
        this._recipient = recipient;
        this._amount = amount;
        this._nineChroniclesTxId = nineChroniclesTxId;
        this._ethereumTransactionHash = ethereumTransactionHash;
    }

    render(): Partial<ChatPostMessageArguments> {
        return {
            text: "wNCG → NCG event occurred.",
            attachments: [
                {
                    author_name: 'Bridge Event',
                    color: "#42f5aa",
                    fields: [
                        {
                            title: "9c network transaction",
                            value: this.toExplorerUrl(this._nineChroniclesTxId),
                        },
                        {
                            title: "Ethereum network transaction",
                            value: this.toEtherscanUrl(this._ethereumTransactionHash),
                        },
                        {
                            title: "sender (Ethereum)",
                            value: this._sender,
                        },
                        {
                            title: "recipient (NineChronicles)",
                            value: this._recipient,
                        },
                        {
                            title: "amount",
                            value: this._amount
                        }
                    ],
                    fallback: `wNCG ${this._sender} → NCG ${this._recipient}`
                }
            ]
        }
    }
}
