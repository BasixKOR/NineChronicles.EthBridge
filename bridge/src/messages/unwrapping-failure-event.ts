import { ChatPostMessageArguments } from "@slack/web-api";
import { combineUrl, Message } from ".";
import { TxId } from "../types/txid";
import { Address } from "../types/address";
import { ForceOmit } from "../types/force-omit";

export class UnwrappingFailureEvent implements Message {
    private readonly _url: string;
    private readonly _sender: Address;
    private readonly _recipient: Address;
    private readonly _txId: TxId;
    private readonly _amount: string;
    private readonly _error: string;

    constructor(
        url: string,
        sender: Address,
        recipient: Address,
        amount: string,
        txId: TxId,
        error: string) {
        this._url = url;
        this._sender = sender;
        this._recipient = recipient;
        this._amount = amount;
        this._txId = txId;
        this._error = error;
    }

    render(): ForceOmit<Partial<ChatPostMessageArguments>, "channel"> {
        return {
            text: "wNCG → NCG event failed.",
            attachments: [
                {
                    author_name: 'Bridge Error',
                    color: "#ff0033",
                    fields: [
                        {
                            title: "Ethereum transaction",
                            value: combineUrl(this._url, `/tx/${this._txId}`),
                        },
                        {
                            title: "sender (Ethereum)",
                            value: this._sender,
                        },
                        {
                            title: "recipient (NineChronicles)",
                            value: this._recipient,
                        },
                        {
                            title: "amount",
                            value: this._amount
                        },
                        {
                            title: "error",
                            value: this._error
                        }
                    ],
                    fallback: `wNCG ${this._sender} → NCG ${this._recipient} failed`
                }
            ]
        }
    }
}
