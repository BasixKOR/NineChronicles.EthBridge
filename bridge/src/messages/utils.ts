import { join } from "path";
import { URL } from "url";

export function combineNcExplorerUrl(
    explorerUrl: string,
    ncscanUrl: string | undefined,
    useNcscan: boolean,
    txId: string
) {
    if (useNcscan) {
        if (ncscanUrl === undefined) {
            throw new Error("ncscanUrl is undefined");
        }

        return combineUrl(ncscanUrl, `/tx/${txId}`);
    } else {
        return combineUrl(explorerUrl, `/transaction?${txId}`);
    }
}

export function combineUrl(base: string, addition: string): string {
    const [path, query] = addition.split("?");
    const url = new URL(base);
    url.pathname = join(url.pathname, path);
    url.search = `?${query || ""}`; // FIXME: remove ? when not required.
    return url.toString();
}
