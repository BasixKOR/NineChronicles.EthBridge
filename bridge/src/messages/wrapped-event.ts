import { ChatPostMessageArguments } from "@slack/web-api";
import { WrappingEvent } from "./wrapping-event";
import { Address } from "../types/address";
import { TxId } from "../types/txid";
import Decimal from "decimal.js";
import { ForceOmit } from "../types/force-omit";

export class WrappedEvent extends WrappingEvent {
    private readonly _sender: Address;
    private readonly _recipient: Address;
    private readonly _nineChroniclesTxId: TxId;
    private readonly _ethereumTransactionHash: string;
    private readonly _fee: Decimal;
    private readonly _exchangeAmount: string;
    private readonly _refundAmount: string | null;
    private readonly _refundTxId: string | null;

    constructor(
        explorerUrl: string,
        ncscanUrl: string | undefined,
        useNcscan: boolean,
        etherscanUrl: string,
        sender: Address,
        recipient: Address,
        exchangeAmount: string,
        nineChroniclesTxId: TxId,
        ethereumTransactionHash: string,
        fee: Decimal,
        refundAmount: string | null,
        refundTxId: TxId | null
    ) {
        super(explorerUrl, ncscanUrl, useNcscan, etherscanUrl);

        this._sender = sender;
        this._recipient = recipient;
        this._exchangeAmount = exchangeAmount;
        this._nineChroniclesTxId = nineChroniclesTxId;
        this._ethereumTransactionHash = ethereumTransactionHash;
        this._fee = fee;
        this._refundAmount = refundAmount;
        this._refundTxId = refundTxId;
    }

    render(): ForceOmit<Partial<ChatPostMessageArguments>, "channel"> {
        const refundFields =
            this._refundAmount !== null && this._refundTxId !== null
                ? [
                      {
                          title: "refund amount",
                          value: this._refundAmount,
                      },
                      {
                          title: "refund transaction",
                          value: this.toExplorerUrl(this._refundTxId),
                      },
                  ]
                : [];
        return {
            text: "NCG → wNCG event occurred.",
            attachments: [
                {
                    author_name: "Bridge Event",
                    color: "#42f5aa",
                    fields: [
                        {
                            title: "9c network transaction",
                            value: this.toExplorerUrl(this._nineChroniclesTxId),
                        },
                        {
                            title: "Ethereum network transaction",
                            value: this.toEtherscanUrl(
                                this._ethereumTransactionHash
                            ),
                        },
                        {
                            title: "sender (NineChronicles)",
                            value: this._sender,
                        },
                        {
                            title: "recipient (Ethereum)",
                            value: this._recipient,
                        },
                        {
                            title: "amount",
                            value: this._exchangeAmount,
                        },
                        {
                            title: "fee",
                            value: this._fee.toString(),
                        },
                        ...refundFields,
                    ],
                    fallback: `NCG ${this._sender} → wNCG ${this._recipient}`,
                },
            ],
        };
    }
}
