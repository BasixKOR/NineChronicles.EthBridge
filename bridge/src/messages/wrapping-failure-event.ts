import { ChatPostMessageArguments } from "@slack/web-api";
import { combineUrl, Message } from ".";
import { TxId } from "../types/txid";
import { Address } from "../types/address";
import { ForceOmit } from "../types/force-omit";

export class WrappingFailureEvent implements Message {
    private readonly _url: string;
    private readonly _sender: Address;
    private readonly _recipient: Address;
    private readonly _txId: TxId;
    private readonly _amount: string;
    private readonly _error: string;

    constructor(
        url: string,
        sender: Address,
        recipient: Address,
        amount: string,
        txId: TxId,
        error: string
    ) {
        this._url = url;
        this._sender = sender;
        this._recipient = recipient;
        this._amount = amount;
        this._txId = txId;
        this._error = error;
    }

    render(): ForceOmit<Partial<ChatPostMessageArguments>, "channel"> {
        return {
            text: "NCG → wNCG event failed.",
            attachments: [
                {
                    author_name: "Bridge Error",
                    color: "#ff0033",
                    fields: [
                        {
                            title: "9c network transaction",
                            value: combineUrl(
                                this._url,
                                `/transaction?${this._txId}`
                            ),
                        },
                        {
                            title: "sender (NineChronicles)",
                            value: this._sender,
                        },
                        {
                            title: "recipient (Ethereum)",
                            value: this._recipient,
                        },
                        {
                            title: "amount",
                            value: this._amount,
                        },
                        {
                            title: "error",
                            value: this._error,
                        },
                    ],
                    fallback: `NCG ${this._sender} → wNCG ${this._recipient} failed`,
                },
            ],
        };
    }
}
