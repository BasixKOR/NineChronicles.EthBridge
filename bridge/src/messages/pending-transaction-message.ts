import { ChatPostMessageArguments } from "@slack/web-api";
import { Message } from ".";
import { ForceOmit } from "../types/force-omit";
import { ExchangeHistory } from "../interfaces/exchange-history-store";
import { combineUrl } from "./utils";
import { MultiPlanetary } from "../multi-planetary";

export class PendingTransactionMessage implements Message {
    private readonly _ethScanUrl: string;
    private readonly _ncScanUrl: string;
    private readonly _multiPlanetary: MultiPlanetary;

    constructor(
        private readonly transactions: ExchangeHistory[],
        private readonly multiPlanetary: MultiPlanetary,
        ethScanUrl: string = process.env.ETHERSCAN_ROOT_URL ||
            "https://etherscan.io/",
        ncScanUrl: string = process.env.NCSCAN_URL || "https://9cscan.com/"
    ) {
        this._ethScanUrl = ethScanUrl;
        this._ncScanUrl = ncScanUrl;
        this._multiPlanetary = multiPlanetary;
    }

    render(): ForceOmit<Partial<ChatPostMessageArguments>, "channel"> {
        if (this.transactions) {
            console.log("Pending Transactions : ", this.transactions);
            return {
                text: `${this.transactions.length} Pending Transactions Found`,
                attachments: this.transactions.map((tx) => {
                    const titleUrl =
                        tx.network === "ethereum"
                            ? this._ethScanUrl
                            : this._ncScanUrl;
                    const senderLabel =
                        tx.network === "ethereum" ? "ETH" : "9C";
                    const recipientLabel =
                        tx.network === "ethereum" ? "9c" : "ETH";
                    const authorLabel =
                        tx.network === "ethereum" ? "wNCG → NCG" : "NCG → wNCG";
                    return {
                        author_name: `[ETH] ${authorLabel} pending event`,
                        color: "#ff0033",
                        fields: [
                            {
                                title: senderLabel + " transaction",
                                value: combineUrl(titleUrl, `/tx/${tx.tx_id}`),
                            },
                            {
                                title: "Planet Name",
                                value: this._multiPlanetary.getRequestPlanetName(
                                    tx.recipient
                                ),
                            },
                            {
                                title: `Sender(${senderLabel})`,
                                value: tx.sender,
                            },
                            {
                                title: `Recipient(${recipientLabel})`,
                                value: tx.recipient,
                            },
                            {
                                title: "Amount",
                                value: tx.amount.toString(),
                            },
                            {
                                title: "Timestamp",
                                value: tx.timestamp,
                            },
                        ],
                    };
                }),
            };
        }

        return {
            text: "No pending transactions",
            attachments: [
                {
                    author_name: "ETH Bridge Restarted",
                    color: "#ffcc00",
                },
            ],
        };
    }
}
