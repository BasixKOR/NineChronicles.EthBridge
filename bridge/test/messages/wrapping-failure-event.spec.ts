import { WrappingFailureEvent } from "../../src/messages/wrapping-failure-event";

describe("WrappingFailureEvent", () => {
    describe("render", () => {
        it("snapshot", () => {
            const EXPLORER_URL = "https://explorer.libplanet.io/9c-internal";
            const NCSCAN_URL = "https://9cscan.com";
            const USE_NCSCAN_URL = false;
            const SENDER = "0xCbfC996ad185c61a031f40CeeE80a055e6D83005";
            const RECIPIENT = "0xDac65eCE9CB3E7a538773e08DE31F973233F064f";
            const AMOUNT = "100";
            const NINE_CHRONICLES_TX_ID =
                "3409cdbaa24ec6f7c8d2c0f636325a2b2e9611e5e6df5c593cfcd299860d8043";
            const failureSubscribers = "@gamefi-be";
            expect(
                new WrappingFailureEvent(
                    EXPLORER_URL,
                    NCSCAN_URL,
                    USE_NCSCAN_URL,
                    SENDER,
                    RECIPIENT,
                    AMOUNT,
                    NINE_CHRONICLES_TX_ID,
                    String(new Error("error")),
                    failureSubscribers
                ).render()
            ).toMatchSnapshot();
        });
    });
});
