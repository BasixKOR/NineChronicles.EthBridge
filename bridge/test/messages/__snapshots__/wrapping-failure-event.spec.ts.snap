// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WrappingFailureEvent render snapshot 1`] = `
Object {
  "attachments": Array [
    Object {
      "author_name": "Bridge Error",
      "color": "#ff0033",
      "fallback": "NCG 0xCbfC996ad185c61a031f40CeeE80a055e6D83005 → wNCG 0xDac65eCE9CB3E7a538773e08DE31F973233F064f failed",
      "fields": Array [
        Object {
          "title": "9c network transaction",
          "value": "https://explorer.libplanet.io/9c-internal/transaction?3409cdbaa24ec6f7c8d2c0f636325a2b2e9611e5e6df5c593cfcd299860d8043",
        },
        Object {
          "title": "sender (NineChronicles)",
          "value": "0xCbfC996ad185c61a031f40CeeE80a055e6D83005",
        },
        Object {
          "title": "recipient (Ethereum)",
          "value": "0xDac65eCE9CB3E7a538773e08DE31F973233F064f",
        },
        Object {
          "title": "amount",
          "value": "100",
        },
        Object {
          "title": "error",
          "value": "Error: error",
        },
      ],
    },
  ],
  "text": "NCG → wNCG event failed. @gamefi-be",
}
`;
